mlr-org/gallery/optimization/2022-12-09-hotstart/index.qmd

```diff
@@ -0,0 +1,191 @@
+If another random forest with 500 trees but with the same hyperparameter configuration has to be trained, the model with 1000 trees is copied and only 500 trees are used for prediction.
+
+We load the `r ref("mlr_learners_classif.ranger", "ranger learner")` and set the search space from the @bischl_hyperparameter_2021 article.
+
+```{r}
+#| message: false
+
+library(mlr3verse)
+
+learner = lrn("classif.ranger",
+  mtry.ratio      = to_tune(0, 1),
+  replace         = to_tune(),
+  sample.fraction = to_tune(1e-1, 1),
+  num.trees       = to_tune(1, 2000)
+)
+```
+
+We activate hotstarting with the `allow_hotstart` option.
+When running a grid search with hotstarting, the grid is sorted by the hot start parameter.
+This means the models with 2000 trees are trained first.
+The models with less than 2000 trees hot start on the 2000 trees models which allows the training to be completed immediately.
+
+```{r}
+#| eval: false
+
+instance = tune(
+  method = tnr("grid_search", resolution = 5, batch_size = 5),
+  task = tsk("spam"),
+  learner = learner,
+  resampling = rsmp("holdout"),
+  measure = msr("classif.ce"),
+  allow_hotstart = TRUE
+)
+```
+
+For comparison, we perform the same tuning without hotstarting.
+
+```{r}
+#| eval: false
+
+instance_2 = tune(
+  method = tnr("grid_search", resolution = 5, batch_size = 5),
+  task = tsk("spam"),
+  learner = learner,
+  resampling = rsmp("holdout"),
+  measure = msr("classif.ce"),
+  allow_hotstart = FALSE
+)
+```
+
+We plot the time of completion of each batch (see @fig-time-batch).
+Each batch includes 5 configurations.
+We can see that tuning with hotstarting is slower at first.
+As soon as all models are fitted with 2000 trees, the tuning runs much faster and overtakes the tuning without hotstarting.
+
+```{r}
+#| echo: false
+#| fig-cap: Time of completion of each batch with and without hotstarting.
+#| label: fig-time-batch
+
+library(ggplot2)
+
+data = readRDS("data.rds")
+data_2 = readRDS("data_2.rds")
+
+data[, runtime := as.numeric(difftime(timestamp, data$timestamp[1]))]
+data_2[, runtime := as.numeric(difftime(timestamp, data_2$timestamp[1]))]
+
+data_3 = rbindlist(list(data, data_2), idcol = "id")
+data_3[, id := factor(id)]
+
+ggplot(data_3, aes(x = runtime, y = batch_nr, group = id)) +
+  geom_line(aes(color = id)) +
+  geom_point(aes(color = id)) +
+  scale_colour_viridis_d(name = "hotstarting", labels = c("Activated", "Deactivated"), end = 0.8) +
+  ylab("Batch") +
+  xlab("Time of Completion [seconds]") +
+  theme_minimal()
+```
+
+# Forward Hotstarting
+
+Forward hotstarting is currently only supported by XGBoost.
+However, we have observed that hotstarting only provides a speed advantage for very large datasets and models with more than 5000 boosting rounds.
+The reason is that copying the models from the main process to the workers is a major bottleneck.
+The parallelization package future copies the models sequentially to the workers.
```

Comment:
Also diese Details sollten finde ich in eine foldable callout box (siehe hier https://quarto.org/docs/authoring/callouts.html), weil für die meisten ist das zu technisch.